trinity/rpc/main.py

```diff
@@ -135,11 +139,11 @@ def _get_result(self,
         else:
             return result, None
 
-    def execute(self, request: Dict[str, Any]) -> str:
+    async def execute(self, request: Dict[str, Any]) -> str:
```

Comment:
It really makes me uneasy how async spreads like a virus. Now any package that wants to use RPC must be async also. It also makes REPL interaction a PITA.

Imagine if the file object returned by `open()` _only_ had async methods. That would be a net negative, right?

I'm not going to advocate to change this back, but... my spider sense is tingling that we will have to do more work in the future to either unravel the mess or support multiple entry points (async and not async).

Reply:
I kinda feel your pain but here's my personal perspective on it.

There's a difference between *event loop based async I/O* in general and `async` / `await` as a pattern for consumption of the former. I feel your pain with `async` / `await` as a pattern but I don't think there is a way to get around *event loop based async I/O* in general.

I think event loop based async I/O is the best way to build heavy I/O dependent software like this especially in a language like Python. Because what else could we do?

1. If we just go synchronous it basically means all I/O is going to run sequentially -> slow

2. Threading out doesn't help much because of the GIL.

Hence, I think that the pattern of event loop based async I/O is fine by itself.

It is the pattern that browsers (as well as nodejs) use since forever and pretty much every browser API (apart from `prompt` which really locks the entire screen) is async because everything else will ruin the user experience. If you write blocking code in the browser, the browser will tell you that "this script is "hanging" and asks you to kill it".

However, **how** we deal with event loop based async I/O is a different story and for most of my life Futures (or Promises for that matter) as well as Observables have been my goto  answer to deal with it. Now I'm slowly adapting to `async` / `await` and part of me likes it and part of me doesn't like it.

However, in a language such as Python working with Futures doesn't feel very welcoming either (`future.add_done_callback(some_fn_that_I_wish_I_could_write_as_a_lambda_but_couldnt`). I guess Observables (RxPy) make it a little nicer (I'm biased I worked with Rx and contributed to various parts since 2010) but still suffers a bit from the fact that lambdas just can't reveal their full potential in Python.

But as I said, I hate the fact that `async` methods spread like a virus as much as you do. That's also why in `lahja` I made it so that you can consume events both ways. Using `async` or callback based.

Because the latter has its place e.g. in situations where you want to consume events without having to `await` it and forcing the function that consumes it to be `async` as well. For instance, we use that in `main.py`.

https://github.com/ethereum/py-evm/blob/dd304cc4ace18116c8f9fb4bbf9161319c9cb4c3/trinity/main.py#L245-L254